### PR TITLE
fix #707 adding post route for /feeds

### LINKED ISSUE #707: Add a POST route to allow a user to add a feed
**What would you like to be added**:
Add a POST route to allow a user to add feeds

**Why would you like this to be added**:
Once #706 goes in we need to allow users to enter their feeds so we can add them to Redis

### PR DESCRIPTION
**Pull request by Grommers00**
## Issue This PR Addresses

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description

This will allow us to connect our front end and back end to post the data to add the feeds to the master list. 

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [ ] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/backend/web/routes/feeds.js [hunk 1]
```diff
@@ -51,4 +51,26 @@ feeds.get('/:id', async (req, res) => {
   }
 });
 
+feeds.post('/', async (req, res) => {
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: This route should get `protect` middleware: you can't do it unless you're logged in.

File: src/backend/web/routes/feeds.js [hunk 2]
```diff
@@ -51,4 +51,26 @@ feeds.get('/:id', async (req, res) => {
...
+  const feedData = req.body;
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: We need to validate this data, since we require that an `author` and `url` get passed.  If either is missing, you need to send back an HTTP 400, indicating that the request is invalid, and that it's missing data.

File: src/backend/web/routes/feeds.js [hunk 3]
```diff
@@ -51,4 +51,26 @@ feeds.get('/:id', async (req, res) => {
...
+
+  try {
+    const feed = await Feed.create(feedData);
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: Let's check if this feed already exists in our system before we try to add it again:

```js
if(await Feed.byUrl(feedData.url)) {
  res.status(409).json({message: `Feed for url ${feedData.url} already exists.`});
  return;
}
```

File: src/backend/web/routes/feeds.js [hunk 4]
```diff
@@ -51,4 +51,26 @@ feeds.get('/:id', async (req, res) => {
...
+
+    // If the object we get back is empty, use 404
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: This "404" block can be removed.  We'll deal with the case of the Redis call failing in your `catch` below.

File: src/backend/web/routes/feeds.js [hunk 5]
```diff
@@ -51,4 +51,26 @@ feeds.get('/:id', async (req, res) => {
...
+    if (!feed) {
+      res.status(404).json({
+        message: `Feed not created for url ${feedData.url}`,
+      });
+    } else {
+      res.json(`${feed.url} has successfully been added.`);
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: When you create a new record, you need to send back a 201, and give them enough info to get the url they need for the record.

```js
res.status(201).json({message: `Feed was successfully added.`, id: feed, url: `/feeds/${feed}` });
```

**Comment by humphd** [comment]
This also needs some tests.

File: src/backend/web/routes/feeds.js (hunk 2 above)
Comment (Grommers00) [reply src/backend/web/routes/feeds.js]: Just wondering though - if it's a protected route, we will have the author correct? 

So, do I just need to check that there is a URL?

Comment (humphd) [reply src/backend/web/routes/feeds.js]: We'll have a username, but not necessarily a real, fullname.  Consider the comment you left in Julia's design issue:

- "humphd"
- vs. "David Humphrey"

The author isn't a user, but a display name.  Later, we might consolidate this somehow, but we have a ton of feeds that don't have associated users, so we can't currently rely on that mapping.

File: src/backend/web/routes/feeds.js [hunk 6]
```diff
@@ -51,4 +52,26 @@ feeds.get('/:id', async (req, res) => {
   }
 });
 
+feeds.post('/', protect, async (req, res) => {
+  const feedData = req.body;
+  if (!feedData.url && !feedData.author) {
+    res.status(400).json({ message: `URL and Author must be submitted` });
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: You need to `return` after this so it doesn't keep processing this request.

File: src/backend/web/routes/feeds.js [hunk 7]
```diff
@@ -51,4 +52,26 @@ feeds.get('/:id', async (req, res) => {
...
+  }
+  if (await Feed.byUrl(feedData.url)) {
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: Throw this `if` block inside your `try/catch`, since `Feed.byUrl()` can throw.

File: src/backend/web/routes/feeds.js [hunk 8]
```diff
@@ -51,4 +52,26 @@ feeds.get('/:id', async (req, res) => {
...
+    res.status(409).json({ message: `Feed for url ${feedData.url} already exists.` });
+    return;
+  }
+  try {
+    const feed = await Feed.create(feedData);
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: Since what you get back from `Feed.create` is an `id` for your feed, let's call it that instead of `feed`: `const id = ...`

File: src/backend/web/routes/feeds.js [hunk 9]
```diff
@@ -51,4 +52,26 @@ feeds.get('/:id', async (req, res) => {
...
+    res
+      .status(201)
+      .json({ message: `Feed was successfully added.`, id: feed, url: `/feeds/${feed}` });
+  } catch (err) {
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: Can you use `error` vs. `err` please, it makes it consistent with the rest of our code.

File: src/backend/web/routes/feeds.js [hunk 10]
```diff
@@ -51,4 +52,26 @@ feeds.get('/:id', async (req, res) => {
   }
 });
 
+feeds.post('/', protect, async (req, res) => {
+  const feedData = req.body;
+  if (!feedData.url && !feedData.author) {
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: This won't cover every case (i.e., this will only get triggered if both are missing, but not one or the other).  You probably want to say, `if(!(feedData.url && feedData.author)) {`

File: src/backend/web/routes/feeds.js (hunk 9 above)
Comment (Grommers00) [reply src/backend/web/routes/feeds.js]: Also fixing the other get routes in the file.

File: src/backend/web/routes/feeds.js (hunk 10 above)
Comment (Grommers00) [reply src/backend/web/routes/feeds.js]: Updated.

File: test/feeds.test.js [hunk 11]
```diff
@@ -62,3 +62,65 @@ describe('test /feeds/:id responses', () => {
     expect(res.body instanceof Array).toBe(false);
   });
 });
+
+describe('test POST /feeds endpoint', () => {
+  const createdItems = 5;
+  const feedData = {
+    author: 'foo',
+    url: 'http://telescope200.cdot.systems',
```
Comment (Silvyre) [inline test/feeds.test.js]: I'm curious about this choice of URL; does this test suite only require a random/broken URL?

File: src/backend/web/routes/feeds.js [hunk 12]
```diff
@@ -2,6 +2,7 @@ const express = require('express');
 const Feed = require('../../data/feed');
 const { getFeeds } = require('../../utils/storage');
 const { logger } = require('../../utils/logger');
+// const { protect } = require('../authentication');
```
Comment (manekenpix) [inline src/backend/web/routes/feeds.js]: I think you can remove this

File: test/feeds.test.js [hunk 13]
```diff
@@ -62,3 +62,65 @@ describe('test /feeds/:id responses', () => {
...
+  it('url already in the feed list', async () => {
+    feedData.url = 'http://telescope3.cdot.systems';
+
+    const res = await request(app)
+      .post('/feeds')
+      .send(feedData)
+      .set('Accept', 'application/json');
+    expect(res.status).toEqual(409);
+  });
+});
```
Comment (manekenpix) [inline test/feeds.test.js]: What's the difference between these 2?

Comment (Grommers00) [reply test/feeds.test.js]: Oops.

File: test/feeds.test.js [hunk 14]
```diff
@@ -62,3 +62,65 @@ describe('test /feeds/:id responses', () => {
...
+  it('responds with json', async () => {
+    const res = await request(app)
+      .post('/feeds')
+      .send(feedData)
+      .set('Accept', 'application/json');
+    expect(res.status).toEqual(201);
+  });
+
+  it('no author being sent', async () => {
+    feedData.author = null;
```
Comment (Silvyre) [inline test/feeds.test.js]: Since, to my knowledge, Jest [test methods](https://jestjs.io/docs/en/api#testname-fn-timeout) run in parallel, modifying a shared resource like this might lead to a race condition.

File: src/backend/web/routes/feeds.js (hunk 12 above)
Comment (Grommers00) [reply src/backend/web/routes/feeds.js]: Once David adds in his code protect will be added back in. @humphd didn't know if I should or just comment in for now as a reminder to add it

File: test/feeds.test.js [hunk 15]
```diff
@@ -62,3 +62,65 @@ describe('test /feeds/:id responses', () => {
...
+    author: 'foo',
+    url: 'http://telescope200.cdot.systems',
+  };
+
+  // Array of feeds
+  const feeds = [...Array(createdItems).keys()].map(item => {
+    return new Feed('foo', `http://telescope${item}.cdot.systems`);
+  });
+
+  beforeAll(() => Promise.all(feeds.map(feed => feed.save())));
```
Comment (manekenpix) [inline test/feeds.test.js]: If you're going to manually add feeds in every test (using POST) you might not need this

File: test/feeds.test.js (hunk 11 above)
Comment (Grommers00) [reply test/feeds.test.js]: So this is where I was having issues, if I used 6 here. It would throw a 409 error. Which showed as a duplicate. Yet I was only creating 5? So it confused me.

File: test/feeds.test.js (hunk 14 above)
Comment (Grommers00) [reply test/feeds.test.js]: That makes more sense why I had to use the 200.

File: test/feeds.test.js (hunk 15 above)
Comment (Grommers00) [reply test/feeds.test.js]: True. But do tests run in parallel or in order? If in order I don't need this. In parallel I would.

**Comment by Grommers00** [comment]
> I think you can include your tests in the `describe` that we already have so you won't have to repeat code ([DRY vs WET](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself#DRY_vs_WET_solutions))

Not sure about this because a Post route is different than a GET route. Unless you were referring to something different?

File: test/feeds.test.js (hunk 11 above)
Comment (Grommers00) [reply test/feeds.test.js]: And it just requires a URL that hasn't been added previously.

File: test/feeds.test.js (hunk 15 above)
Comment (manekenpix) [reply test/feeds.test.js]: This part only adds feeds using `feed.save()`, which is not very useful to you since you want to do `POST` and test `json`, `url`, `author`. I think maybe creating 3 Feed objects (one is ok, one without author, one without url) and using them in the specific test could be all you need.

**Comment by manekenpix** [comment]
> Not sure about this because a Post route is different than a GET route. Unless you were referring to something different?

You're right. If you end up not using `beforeAll` maybe it's better to have your tests in a different set.

**Comment by humphd** [comment]
Some really good ideas in here.  I'll just add/answer a few more:

- consider each test as an individual "unit", and don't rely on anything external to that test if possible.  For example, if you need some data to use, create it in that test.  I usually do this by having functions that return data, and call them over and over again in different tests vs. creating variables.  Variables are what make tests fail.
- Tests can and will run in any order.  As such, don't have one test depend on state from another (e.g., don't add something in one test, and try to read it in the next test).
- Tests might not all get run.  A dev might run a single test, so don't rely on anything in other tests.
- If you require the world to be in a certain state for your test, do that in the test itself.  If all your tests need that, do `beforeAll()` or `beforeEach()`, but prefer doing things in the test if possible so you know they happened in the order you need them to, and nothing else changed their state.

In summary: don't share data between tests, don't rely on one test in another, don't rely on test order.

File: test/feeds.test.js [hunk 16]
```diff
@@ -62,3 +62,54 @@ describe('test /feeds/:id responses', () => {
     expect(res.body instanceof Array).toBe(false);
   });
 });
+
+describe('test POST /feeds endpoint', () => {
+  it('responds with json', async () => {
+    const feedData1 = {
```
Comment (humphd) [inline test/feeds.test.js]: Minor nit: all these variables can be named `feedData` without the 1, 2, 3 suffix, since they are local to the function.

File: src/backend/web/routes/feeds.js [hunk 17]
```diff
@@ -34,21 +35,42 @@ feeds.get('/:id', async (req, res) => {
 
   try {
     const feed = await Feed.byId(id);
-
-    // If the object we get back is empty, use 404
```
Comment (humphd) [inline src/backend/web/routes/feeds.js]: Not sure why you removed these lines.  Not critical, but I'd avoid changing things not related to your code.

File: test/feeds.test.js [hunk 18]
```diff
@@ -62,3 +62,55 @@ describe('test /feeds/:id responses', () => {
...
+    const feedData = {
+      author: 'foo',
+      url: 'http://telescope200.cdot.systems',
+    };
+    const res = await request(app)
+      .post('/feeds')
+      .send(feedData)
+      .set('Accept', 'application/json');
+    expect(res.status).toEqual(201);
+  });
```
Comment (humphd) [inline test/feeds.test.js]: Let's modify this test a bit.  I care less about it responding with JSON, and more that it gives back a usable id.  After your code above, do this:

```js
// See if the feed was created as expected.
const id = res.body.id;
const feed = await Feed.byId(id);
expect(feed.author).toEqual(feedData.author);
expect(feed.url).toEqual(feedData.url);
```

This code is untested, but hopefully it gives you enough.

File: src/backend/web/routes/feeds.js (hunk 17 above)
Comment (Grommers00) [reply src/backend/web/routes/feeds.js]: Misclick :)
